### PR TITLE
Fix infinite loop with rebinding after-save-hook (fixes #27)

### LINKED ISSUE #27: Global mode conflicts with php-mode
emacsrc:

~~~
(require 'php-mode)
(require 'apheleia)

(add-to-list 'apheleia-formatters '(cat "cat"))
(add-to-list 'apheleia-mode-alist '(php-mode . cat))

(apheleia-global-mode +1)

(setq debug-on-message "while locally")
~~~

test.php:

~~~
Something here
~~~

Run this (you should have apheleia, php-mode installed with straight, or clone the repos elsewhere and change the paths below):

~~~
emacs -Q -L ~/.emacs.d/straight/repos/apheleia -L ~/.emacs.d/straight/repos/php-mode/lisp -l emacsrc test.php
~~~

After changing something and saving, you get:

~~~
Debugger entered--Lisp error: "Making after-save-hook buffer-local while locally ..."
  make-local-variable(after-save-hook)
  add-hook(after-save-hook apheleia--format-after-save nil local)
  (if apheleia-mode (add-hook 'after-save-hook #'apheleia--format-after-save nil 'local) (remove-hook 'after-save-hook #'apheleia--format-after-save 'local))
  (let ((last-message (current-message))) (setq apheleia-mode (if (eq arg 'toggle) (not apheleia-mode) (> (prefix-numeric-value arg) 0))) (if apheleia-mode (add-hook 'after-save-hook #'apheleia--format-after-save nil 'local) (remove-hook 'after-save-hook #'apheleia--format-after-save 'local)) (run-hooks 'apheleia-mode-hook (if apheleia-mode 'apheleia-mode-on-hook 'apheleia-mode-off-hook)) (if (called-interactively-p 'any) (progn nil (if (and (current-message) (not (equal last-message (current-message)))) nil (let ((local " in current buffer")) (message "Apheleia mode %sabled%s" (if apheleia-mode "en" "dis") local))))))
  apheleia-mode()
  (if apheleia-mode (progn (apheleia-mode -1) (apheleia-mode)) (apheleia-mode))
  (if (eq apheleia-mode-major-mode major-mode) nil (if apheleia-mode (progn (apheleia-mode -1) (apheleia-mode)) (apheleia-mode)))
  (if apheleia-mode-set-explicitly nil (if (eq apheleia-mode-major-mode major-mode) nil (if apheleia-mode (progn (apheleia-mode -1) (apheleia-mode)) (apheleia-mode))))
  (save-current-buffer (set-buffer buf) (if apheleia-mode-set-explicitly nil (if (eq apheleia-mode-major-mode major-mode) nil (if apheleia-mode (progn (apheleia-mode -1) (apheleia-mode)) (apheleia-mode)))) (setq apheleia-mode-major-mode major-mode))
  (progn (save-current-buffer (set-buffer buf) (if apheleia-mode-set-explicitly nil (if (eq apheleia-mode-major-mode major-mode) nil (if apheleia-mode (progn (apheleia-mode -1) (apheleia-mode)) (apheleia-mode)))) (setq apheleia-mode-major-mode major-mode)))
  (if (buffer-live-p buf) (progn (save-current-buffer (set-buffer buf) (if apheleia-mode-set-explicitly nil (if (eq apheleia-mode-major-mode major-mode) nil (if apheleia-mode (progn (apheleia-mode -1) (apheleia-mode)) (apheleia-mode)))) (setq apheleia-mode-major-mode major-mode))))
  (let ((buf (car --dolist-tail--))) (if (buffer-live-p buf) (progn (save-current-buffer (set-buffer buf) (if apheleia-mode-set-explicitly nil (if (eq apheleia-mode-major-mode major-mode) nil (if apheleia-mode (progn ... ...) (apheleia-mode)))) (setq apheleia-mode-major-mode major-mode)))) (setq --dolist-tail-- (cdr --dolist-tail--)))
  (while --dolist-tail-- (let ((buf (car --dolist-tail--))) (if (buffer-live-p buf) (progn (save-current-buffer (set-buffer buf) (if apheleia-mode-set-explicitly nil (if (eq apheleia-mode-major-mode major-mode) nil (if apheleia-mode ... ...))) (setq apheleia-mode-major-mode major-mode)))) (setq --dolist-tail-- (cdr --dolist-tail--))))
  (let ((--dolist-tail-- buffers)) (while --dolist-tail-- (let ((buf (car --dolist-tail--))) (if (buffer-live-p buf) (progn (save-current-buffer (set-buffer buf) (if apheleia-mode-set-explicitly nil (if ... nil ...)) (setq apheleia-mode-major-mode major-mode)))) (setq --dolist-tail-- (cdr --dolist-tail--)))))
  (let ((buffers apheleia-global-mode-buffers)) (setq apheleia-global-mode-buffers nil) (let ((--dolist-tail-- buffers)) (while --dolist-tail-- (let ((buf (car --dolist-tail--))) (if (buffer-live-p buf) (progn (save-current-buffer (set-buffer buf) (if apheleia-mode-set-explicitly nil ...) (setq apheleia-mode-major-mode major-mode)))) (setq --dolist-tail-- (cdr --dolist-tail--))))))
  apheleia-global-mode-enable-in-buffers()
  run-hooks(after-change-major-mode-hook)
  run-mode-hooks(php-mode-hook)
  php-mode()
  funcall(php-mode)
  php-mode-maybe()
  set-auto-mode-0(php-mode-maybe t)
  set-auto-mode(t)
  set-visited-file-name("/tmp/tmp.PViU3Ex3tP/test.php" t)
  write-file("/tmp/tmp.PViU3Ex3tP/test.php")
  (let ((after-save-hook (remq #'apheleia--format-after-save after-save-hook))) (write-file (or filename buffer-file-name)))
  (progn (fset #'message vnew) (let ((after-save-hook (remq #'apheleia--format-after-save after-save-hook))) (write-file (or filename buffer-file-name))))
  (unwind-protect (progn (fset #'message vnew) (let ((after-save-hook (remq #'apheleia--format-after-save after-save-hook))) (write-file (or filename buffer-file-name)))) (fset #'message old))
  (let* ((vnew #'(lambda (format &rest args) (if (equal format "Saving file %s...") nil (apply message format args)))) (old (symbol-function #'message))) (unwind-protect (progn (fset #'message vnew) (let ((after-save-hook (remq #'apheleia--format-after-save after-save-hook))) (write-file (or filename buffer-file-name)))) (fset #'message old)))
  (let ((message (symbol-function #'message))) (let* ((vnew #'(lambda (format &rest args) (if (equal format "Saving file %s...") nil (apply message format args)))) (old (symbol-function #'message))) (unwind-protect (progn (fset #'message vnew) (let ((after-save-hook (remq ... after-save-hook))) (write-file (or filename buffer-file-name)))) (fset #'message old))))
  (progn (fset #'write-region vnew) (let ((message (symbol-function #'message))) (let* ((vnew #'(lambda (format &rest args) (if ... nil ...))) (old (symbol-function #'message))) (unwind-protect (progn (fset #'message vnew) (let ((after-save-hook ...)) (write-file (or filename buffer-file-name)))) (fset #'message old)))))
  (unwind-protect (progn (fset #'write-region vnew) (let ((message (symbol-function #'message))) (let* ((vnew #'(lambda ... ...)) (old (symbol-function #'message))) (unwind-protect (progn (fset #'message vnew) (let (...) (write-file ...))) (fset #'message old))))) (fset #'write-region old))
  (let* ((vnew #'(lambda (start end filename &optional append _visit lockname mustbenew) (apheleia--write-region-silently start end filename append t lockname mustbenew write-region))) (old (symbol-function #'write-region))) (unwind-protect (progn (fset #'write-region vnew) (let ((message (symbol-function #'message))) (let* ((vnew #'...) (old (symbol-function ...))) (unwind-protect (progn (fset ... vnew) (let ... ...)) (fset #'message old))))) (fset #'write-region old)))
  (let ((write-region (symbol-function #'write-region))) (let* ((vnew #'(lambda (start end filename &optional append _visit lockname mustbenew) (apheleia--write-region-silently start end filename append t lockname mustbenew write-region))) (old (symbol-function #'write-region))) (unwind-protect (progn (fset #'write-region vnew) (let ((message (symbol-function ...))) (let* ((vnew ...) (old ...)) (unwind-protect (progn ... ...) (fset ... old))))) (fset #'write-region old))))
  apheleia--write-file-silently("/tmp/tmp.PViU3Ex3tP/test.php")
  (progn (apheleia--write-file-silently buffer-file-name) (run-hooks 'apheleia-post-format-hook))
  (condition-case err (progn (apheleia--write-file-silently buffer-file-name) (run-hooks 'apheleia-post-format-hook)) ((debug error) (message "Apheleia: %s" err) nil))
  (closure ((command "cat") apheleia-mode t) nil (condition-case err (progn (apheleia--write-file-silently buffer-file-name) (run-hooks 'apheleia-post-format-hook)) ((debug error) (message "Apheleia: %s" err) nil)))()
  funcall((closure ((command "cat") apheleia-mode t) nil (condition-case err (progn (apheleia--write-file-silently buffer-file-name) (run-hooks 'apheleia-post-format-hook)) ((debug error) (message "Apheleia: %s" err) nil))))
  (progn (funcall callback))
  (if callback (progn (funcall callback)))
  (progn (apheleia--apply-rcs-patch (current-buffer) patch-buffer) (if callback (progn (funcall callback))))
  (if (equal apheleia--buffer-hash (apheleia--buffer-hash)) (progn (apheleia--apply-rcs-patch (current-buffer) patch-buffer) (if callback (progn (funcall callback)))))
  (save-current-buffer (set-buffer cur-buffer) (if (equal apheleia--buffer-hash (apheleia--buffer-hash)) (progn (apheleia--apply-rcs-patch (current-buffer) patch-buffer) (if callback (progn (funcall callback))))))
  (closure ((formatted-buffer . #<buffer  *apheleia-cat-stdout*>) (cur-buffer . #<buffer test.php>) (callback closure ((command "cat") apheleia-mode t) nil (condition-case err (progn (apheleia--write-file-silently buffer-file-name) (run-hooks 'apheleia-post-format-hook)) ((debug error) (message "Apheleia: %s" err) nil))) (command "cat") t) (patch-buffer) (save-current-buffer (set-buffer cur-buffer) (if (equal apheleia--buffer-hash (apheleia--buffer-hash)) (progn (apheleia--apply-rcs-patch (current-buffer) patch-buffer) (if callback (progn (funcall callback)))))))(#<buffer  *apheleia-diff-stdout*>)
  funcall((closure ((formatted-buffer . #<buffer  *apheleia-cat-stdout*>) (cur-buffer . #<buffer test.php>) (callback closure ((command "cat") apheleia-mode t) nil (condition-case err (progn (apheleia--write-file-silently buffer-file-name) (run-hooks 'apheleia-post-format-hook)) ((debug error) (message "Apheleia: %s" err) nil))) (command "cat") t) (patch-buffer) (save-current-buffer (set-buffer cur-buffer) (if (equal apheleia--buffer-hash (apheleia--buffer-hash)) (progn (apheleia--apply-rcs-patch (current-buffer) patch-buffer) (if callback (progn (funcall callback))))))) #<buffer  *apheleia-diff-stdout*>)
  (progn (funcall callback stdout))
  (if callback (progn (funcall callback stdout)))
  (if (funcall (or exit-status #'(lambda (status) (= 0 status))) (process-exit-status proc)) (if callback (progn (funcall callback stdout))) (message (concat "Failed to run %s: exit status %s " "(see hidden buffer%s)") (car command) (process-exit-status proc) stderr))
  (if (process-live-p proc) nil (save-current-buffer (set-buffer stderr) (if (= 0 (buffer-size)) (progn (insert "[No output received on stderr]\n")))) (if (funcall (or exit-status #'(lambda (status) (= 0 status))) (process-exit-status proc)) (if callback (progn (funcall callback stdout))) (message (concat "Failed to run %s: exit status %s " "(see hidden buffer%s)") (car command) (process-exit-status proc) stderr)))
  (closure ((stderr . #<buffer  *apheleia-diff-stderr*>) (stdout . #<buffer  *apheleia-diff-stdout*>) (name . "diff") (exit-status closure ((new-fname) (old-fname . "/tmp/tmp.PViU3Ex3tP/test.php") (callback closure ((formatted-buffer . #<buffer  *apheleia-cat-stdout*>) (cur-buffer . #<buffer test.php>) (callback closure ... nil ...) (command "cat") t) (patch-buffer) (save-current-buffer (set-buffer cur-buffer) (if ... ...))) (new-buffer . #<buffer  *apheleia-cat-stdout*>) (old-buffer . #<buffer test.php>) t) (status) (memq status '(0 1))) (callback closure ((formatted-buffer . #<buffer  *apheleia-cat-stdout*>) (cur-buffer . #<buffer test.php>) (callback closure ((command "cat") apheleia-mode t) nil (condition-case err (progn ... ...) (... ... nil))) (command "cat") t) (patch-buffer) (save-current-buffer (set-buffer cur-buffer) (if (equal apheleia--buffer-hash (apheleia--buffer-hash)) (progn (apheleia--apply-rcs-patch ... patch-buffer) (if callback ...))))) (stdin . #<buffer  *apheleia-cat-stdout*>) (command "diff" "--rcs" "--" "/tmp/tmp.PViU3Ex3tP/test.php" "-") (--cl-rest-- :command ("diff" "--rcs" "--" "/tmp/tmp.PViU3Ex3tP/test.php" "-") :stdin #<buffer  *apheleia-cat-stdout*> :callback (closure ((formatted-buffer . #<buffer  *apheleia-cat-stdout*>) (cur-buffer . #<buffer test.php>) (callback closure (... apheleia-mode t) nil (condition-case err ... ...)) (command "cat") t) (patch-buffer) (save-current-buffer (set-buffer cur-buffer) (if (equal apheleia--buffer-hash ...) (progn ... ...)))) :exit-status (closure ((new-fname) (old-fname . "/tmp/tmp.PViU3Ex3tP/test.php") (callback closure (... ... ... ... t) (patch-buffer) (save-current-buffer ... ...)) (new-buffer . #<buffer  *apheleia-cat-stdout*>) (old-buffer . #<buffer test.php>) t) (status) (memq status '(0 1)))) t) (proc _event) (if (process-live-p proc) nil (save-current-buffer (set-buffer stderr) (if (= 0 (buffer-size)) (progn (insert "[No output received on stderr]\n")))) (if (funcall (or exit-status #'(lambda ... ...)) (process-exit-status proc)) (if callback (progn (funcall callback stdout))) (message (concat "Failed to run %s: exit status %s " "(see hidden buffer%s)") (car command) (process-exit-status proc) stderr))))(#<process aphelieia-diff> "finished\n")
~~~

This is due to apheleia let-binding after-save-hook while writing the file. Writing the file triggers auto mode, auto mode triggers php mode, php mode triggers the hook installed by apheleia global mode which should add apheleia's after save hook, and that breaks because after-save-hook is locally let bound at that point.

https://github.com/raxod502/apheleia/blob/8a1e68441ca418c2a277d7aef663790f26208dd8/apheleia.el#L351-L354

Instead of binding after-save-hook itself, should we not better bind a flag, eg, `apheleia--format-after-save-in-progress`, to break the infinite loop?

I will submit a PR for your consideration.

### PR DESCRIPTION
**Pull request by apeschar**
See issue #27.

### CODE REVIEW

File: apheleia.el [hunk 1]
```diff
@@ -629,19 +626,27 @@ changes), CALLBACK, if provided, is invoked with no arguments."
...
   "Run code formatter for current buffer if any configured, then save."
-  (when apheleia-mode
-    (when-let ((command (apheleia--get-formatter-command)))
-      (apheleia-format-buffer
-       command
-       (lambda ()
-         (with-demoted-errors "Apheleia: %s"
-           (apheleia--write-file-silently buffer-file-name)
-           (run-hooks 'apheleia-post-format-hook)))))))
+  (when (not apheleia--format-after-save-in-progress)
```
Comment (raxod502) [inline apheleia.el:640]:
```suggestion
  (unless apheleia--format-after-save-in-progress
```

**Review by raxod502 — APPROVED** [review]
:sparkles: This looks great, thank you!

**Comment by terlar** [comment]
Was this the root issue of #18 as well? I will have another look to see if that is the case.

**Comment by terlar** [comment]
Seems not, if I enable apheleia-mode the editor keeps freezing for extended periods

**Pull request merged**